build: detect and abort build when getPath is called outside of any make

marler8997 opened the pull request:
zig's build system now detects when any LazyPath.getPath function is called outside of any make function.  This is done by saving the currently running step to a threadlocal global variable.  An added benefit of this change is we can use this thread local variable to provide better error messages without having to manage/propagate the currently running step throughout the build.

I also have a follow up change in the works that will be able to detect whenever any step tries to use a lazy path without having dependend on it first, but this requires more changes to the API that I'd rather defer to its own PR. With that branch I had found a few errors and included one of the fixes in this commit (see `RemoveDir.zig`).  Also, with this PR and the follow up, we'll have everything we need to be able to implement https://github.com/ziglang/zig/issues/21525 (the ability to only fetch lazy dependencies if they are required by the current set of steps/targets).

andrewrk reviewed (changes requested):
I don't understand why you think thread local data is required to implement this?

I'd rather work on #20981 than this.

marler8997 commented:
To clarify, this change allows the build system to know which step is currently being executed.  By doing this, the build system can verify misconfiguation.  In particular, the goal here is to catch out steps that are using lazy paths which they have no dependency on (either direct or indirect).

This PR along with a branch I'm working on has caught missing dependencies in the std build system (see https://github.com/ziglang/zig/compare/master...marler8997:zig:lazyPathMisuse).  There's one in `RemoveDir.zig`, a few in `CheckObject.zig`. I've also fixed missing dependencies for lazy paths in the past in this PR: https://github.com/ziglang/zig/pull/20475/files.  These missing dependencies never would have made it through CI if we had the verification I've added here and in the branch I linked.

Why is this important? Because implementing #21525 requires that the build system is able to find all "lazy paths" that will be needed during a build so it can find the correct set of required lazy dependencies. Steps today are supposed to declare their dependency on lazy paths via `LazyPath.addStepDependencies` but not all do, so depending on this would be brittle. Adding a way to enforce/check when steps fail to do this makes it viable.

I'm not seeing how https://github.com/ziglang/zig/issues/20981 (running the configure phase in a separate process) fixes this.  The same bugs I listed above would still exist whether or not the configure phase is separated out.

marler8997 commented:
Actually maybe I do see a way that #20981 could solve this?  If we run the configuration in a separate process, then that has alot of implications, some I'm not so clear on.  It means everything in the build graph is "serialized" so maybe we can detect misconfiguration at that time?  Not sure though, there's alot of unknowns as to how that would work/look, but, theoretically both the dependence and usage of lazy paths would be codified in that serialized graph and we could verify it somehow?

andrewrk commented:
Thanks but I'm going to veto this based on usage of threadlocal and that this code will be unnecessary when the build runner is a different process than configure. Path forward is #20981 or #14941.

marler8997 commented:
This PR has a few nice tests...I can see how separating the config/build phase fixes the immediate issue of preventing `getPath` from being called in the configure phase but that's not the main benefit of these changes.  I'm not sure how those issues you linked would detect misconfiguration where `getPath` is called on a `LazyPath` that the step forgot to explicitly dependenncy on? Am I missing something? Can you confirm that those issues would cause the tests I added here to pass? see

[test/standalone/lazypath/example/build.zig](https://github.com/ziglang/zig/pull/21541/files#diff-0240a711b0ad15147e19a72d132eb6e0e9b481bbe3c04ccd3799581049104c3e)

This change fixes the issue for Generated lazy steps, and a follow up change would detect this misconfiguration for all kinds of LazyPaths regardless of whether or not they are generated.

marler8997 commented on lib/std/Build/Step/RemoveDir.zig:
```diff
@@ -20,6 +20,7 @@ pub fn create(owner: *std.Build, doomed_path: LazyPath) *RemoveDir {
         }),
         .doomed_path = doomed_path.dupe(owner),
     };
+    doomed_path.addStepDependencies(&remove_dir.step);
```
FYI, since Andrew has veoted this change we might want to add this bug fix with another PR.  There are more big fixes like this in the planned follow up PR but without my changes they won't be detected, we have to manually find them with my branch and then port them over to upstream.

marler8997 replied:
I think I found one instance where there was more missing lazy path dependencies in `CheckObject.zig`:
https://github.com/ziglang/zig/compare/master...marler8997:zig:lazyPathMisuse#diff-250b661d4bed9463d5fef7e603cd49501d1ffb8b439330ce818984247abb1552
